apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/StorageClientImplTest.cs

```diff
@@ -69,5 +76,402 @@ public void GetEffectiveEncryptionKey_Invalid(EncryptionKey clientkey, Encryptio
             var exception = Assert.Throws<ArgumentException>(() => client.GetEffectiveEncryptionKey(operationKey, kmsKey));
             Assert.Contains(expectedMessageSubstring, exception.Message);
         }
+
+        // Tests for GCS Retry :
+        // - No retry applied
+        // - Retry then success (sometimes with conditions)
+        // - Retry with eventual failure
+
+
+        private static void NoRetryHelper<T>(
+            Func<StorageService, ClientServiceRequest<T>> requestProvider,
+            Action<StorageClient> clientAction,
+            HttpStatusCode statusCode = HttpStatusCode.NotFound)
+        {
+            var service = new FakeStorageService();
+            var client = new StorageClientImpl(service);
+
+            service.ExpectRequest(requestProvider(service), statusCode);
+            Assert.Throws<GoogleApiException>(() => clientAction(client));
+            service.Verify();
+        }
+
+        private static void RetryOnce<T>(Func<StorageService, ClientServiceRequest<T>> requestProvider, Action<StorageClient> clientAction, T response, 
+            HttpStatusCode firstStatusCode = HttpStatusCode.BadGateway)
+        {
+            var service = new FakeStorageService();
+            var client = new StorageClientImpl(service);
+            var request = requestProvider(service);
+
+            service.ExpectRequest(request, firstStatusCode);
+            service.ExpectRequest(request, response);
+            clientAction(client);
+            service.Verify();
+        }
+
+        private static void RetryOnceHelper<T>(Func<StorageService, ClientServiceRequest<T>> requestProvider, Action<StorageClient> clientAction,
+           HttpStatusCode firstStatusCode = HttpStatusCode.BadGateway)
+           where T : new() => RetryOnce(requestProvider, clientAction, new T(), firstStatusCode);
+
+        private static void RetryThenFail<T>(Func<StorageService, ClientServiceRequest<T>> requestProvider, Action<StorageClient> clientAction, T response, 
+            HttpStatusCode errorCode = HttpStatusCode.BadGateway, int retryCount = 3)
+        {
+            var service = new FakeStorageService();
+            var client = new StorageClientImpl(service);
+            var request = requestProvider(service);
+
+            for (int i = 1; i <= retryCount; i++)
+            {
+                service.ExpectRequest(request, errorCode);
+            }
+            Assert.Throws<GoogleApiException>(() => clientAction(client));
+            service.Verify();
+        }
+
+        private static void RetryThenFailHelper<T>(Func<StorageService, ClientServiceRequest<T>> requestProvider, Action<StorageClient> clientAction,
+            HttpStatusCode errorCode = HttpStatusCode.BadGateway, int retryCount = 3)
+            where T : new() => RetryThenFail(requestProvider, clientAction, new T(), errorCode, retryCount);
+
+        // Bucket test cases BEGIN
```

Comment:
While I'm not a massive fan of regions, I think they'd be reasonable here - and allow someone to really easily collapse all the bucket test cases, etc.